src/devices/SenseHat/SenseHatLedMatrix.cs

```diff
@@ -95,9 +138,216 @@ public static int PositionToIndex(int x, int y)
+        {
+            get
+            {
+                return _textBackgroundColor;
+            }
+            set
+            {
+                _textBackgroundColor = value;
+                RenderText();
+            }
+        }
+
+        /// <summary>
+        /// Text background color when rendering text.
+        /// </summary>
+        [Property]
+        public double TextScrollPixelsPerSecond
+        {
+            get
+            {
+                return _textScrollPixelsPerSecond;
+            }
+            set
+            {
+                _textScrollPixelsPerSecond = value;
+                StartOrStopTextScrolling();
+            }
+        }
+
+        /// <summary>
+        /// Text rotation, counterclockwise.
+        /// </summary>
+        [Property]
+        public SenseHatTextRotation TextRotation
+        {
+            get
+            {
+                return _textRotation;
+            }
+            set
+            {
+                _textRotation = value;
+                RenderText();
+            }
+        }
+
+        private void StartOrStopTextScrolling()
+        {
+            var renderMatrix = _textRenderMatrix;
+            int millisecondsPerPixel;
+            if (renderMatrix == null || renderMatrix.Text.Length == 0 || _textScrollPixelsPerSecond <= 0)
+            {
+                millisecondsPerPixel = 0;
+            }
+            else
+            {
+                millisecondsPerPixel = (int)(1000 / _textScrollPixelsPerSecond);
+            }
+
+            if (millisecondsPerPixel <= 0)
+            {
+                StopTextAnimationTimer();
+            }
+            else
+            {
+                // Calculate the number of milliseconds for one pixel shift
+                StartTextAnimationTimer((int)millisecondsPerPixel);
+            }
+        }
+
+        private void StartTextAnimationTimer(int intervalMs)
+        {
+            StopTextAnimationTimer();
+            _textAnimationTimer = new System.Timers.Timer(intervalMs);
+            _textAnimationTimer.Elapsed += TextAnimationTimer_Elapsed;
+            _textAnimationTimer.Start();
+        }
+
+        private void StopTextAnimationTimer()
+        {
+            if (_textAnimationTimer != null)
+            {
+                _textAnimationTimer.Stop();
+                _textAnimationTimer.Elapsed -= TextAnimationTimer_Elapsed;
+                _textAnimationTimer.Dispose();
+                _textAnimationTimer = null;
+            }
+        }
+
+        private void TextAnimationTimer_Elapsed(object? sender, System.Timers.ElapsedEventArgs e)
+        {
+            var renderMatrix = _textRenderMatrix;
+            if (renderMatrix != null && renderMatrix.Text.Length > 1)
+            {
+                renderMatrix.ScrollByOnePixel();
+                RenderText();
+            }
+        }
+
+        private void RenderText()
+        {
+            lock (_lockTextRender)
```

Comment:
that's a bit expensive as a lock, any way to lock less that everything?
If my understanding is correct, the only 2 critical parts is _textRenderMatrix, so you may want to lock and check it  on when calling _textRenderMatrix.IsPixelSet

Reply:
Yes, that was a thorn in my eye also. It was necessary because the text animation timer could kick in mid-frame, and with every `SetPixel` the underlying device (i.e. i2c bus) is accessed. If the device is disposed mid frame, then it will cause an exception. Besides that, frame consistency is an issue.

I have refactored the locking approach by introducing a render state class holding a consistent set of parameters, and by using the `Write` method at the end of `RenderText`. This still requires a lock but it is brief and performance is much better.

You are probably going to ask whether the frame buffer needs to be allocated for each `RenderText`. It could be avoided with an additional lock, but I think it's OK as it is. What do you think?

Reply:
Yes, it's definitely better from my point of view!